src/test/java/org/symphonyoss/symphony/messageml/MessageMLContextTest.java

```diff
@@ -337,6 +340,134 @@ public void testParseMessageMLRadioWithLabels()
+            "</messageML>";
+
+    context.parseMessageML(message, "", MessageML.MESSAGEML_VERSION);
+    BiContext biContext = context.getBiContext();
+
+    assertEquals(3, biContext.getItems().size());
+    assertEquals("Form", biContext.getItems().get(0).getName());
+    assertEquals("TextField", biContext.getItems().get(1).getName());
+    assertEquals("Button", biContext.getItems().get(2).getName());
+  }
+
+  @Test
+  public void testGetBiContextWithSameElementsInForm() throws InvalidInputException, IOException, ProcessingException {
+    final String message =
+            "<messageML>" +
+            "    <form id=\"all-elements\">" +
+            "        <button name=\"example-button2\" class=\"primary\" type=\"action\" title =\"This is a title\">Button Text</button>" +
+            "        <button name=\"example-button2\" class=\"primary\" type=\"action\">Button Text</button>" +
+            "    </form>" +
+            "</messageML>";
+
+    context.parseMessageML(message, "", MessageML.MESSAGEML_VERSION);
+    BiContext biContext = context.getBiContext();
+
+    assertEquals(3, biContext.getItems().size());
+    assertEquals("Form", biContext.getItems().get(0).getName());
+    assertEquals("Button", biContext.getItems().get(1).getName());
+    assertEquals("Button", biContext.getItems().get(2).getName());
+  }
+
+  @Test
+  public void testGetBiContextDuplicateSimpleTags() throws InvalidInputException, IOException, ProcessingException {
+    final String message =
+            "<messageML>" +
+              "<a href=\"https://hello.org\">Hello world!</a>" +
+              "<a href=\"http:google.com\">Google!</a>" +
+              "<emoji shortcode=\"smiley\"/>" +
+            "</messageML>";
+
+    Map<String, String> expectedLinkAttrs = new HashMap<>();
+    expectedLinkAttrs.put("href", "2");
+    Map<String, String> expectedEmojisAttrs = new HashMap<>();
+    expectedEmojisAttrs.put("emoji", "1");
+
+    context.parseMessageML(message, "", MessageML.MESSAGEML_VERSION);
+    BiContext biContext = context.getBiContext();
+
+    List<BiItem> biItems = biContext.getItems();
+    assertEquals(2, biItems.size());
+
+    BiItem linkItem = biItems.get(0);
+    assertEquals("Link", linkItem.getName());
+    assertEquals(expectedLinkAttrs, linkItem.getAttributes());
+
+    BiItem emojiItem = biItems.get(1);
+    assertEquals("Emoji", biItems.get(1).getName());
+    assertEquals(expectedEmojisAttrs, emojiItem.getAttributes());
+  }
+
+  @Test
+  public void testGetBiContextWithMultipleHeaders() throws InvalidInputException, IOException, ProcessingException {
+    final String message =
+            "<messageML>" +
+              "<h1>Hello world!</h1>" +
+              "<h2>Hello world!</h2>" +
+              "<h2>Hello world!</h2>" +
+              "<h5>Hello world!</h5>" +
+            "</messageML>";
+
+    Map<String, String> expectedHeadersAttributes = new HashMap<>();
+    expectedHeadersAttributes.put("h1", "1");
```

Comment:
Now I'm wondering if `Map<String, String>` is the container for attributes, we might prefer `Map<String, Object>` to preserve the value type?

Reply:
Done in this commit https://github.com/symphonyoss/messageml-utils/pull/259/commits/fe6d84b9c99ccb08e32ea8197ec2ed2cd4870399

Reply:
Great, looks more natural this way